tests/aws/services/stepfunctions/v2/test_sfn_api.py

```diff
@@ -315,6 +317,102 @@ def test_list_sms(
         sfn_snapshot.match("lst_resp_del_filter", lst_resp_filter)
 
+    @markers.aws.validated
+    def test_list_sms_pagination(
+        self, create_iam_role_for_sfn, create_state_machine, sfn_snapshot, aws_client
+    ):
+        snf_role_arn = create_iam_role_for_sfn()
+
+        sfn_snapshot.add_transformer(RegexTransformer(snf_role_arn, "snf_role_arn"))
+
+        definition = BaseTemplate.load_sfn_template(BaseTemplate.BASE_PASS_RESULT)
+        definition_str = json.dumps(definition)
+
+        sm_names = [f"statemachine_{i}_{short_uid()}" for i in range(13)]
+        state_machine_arns = list()
+
+        for i, sm_name in enumerate(sm_names):
+            creation_resp = create_state_machine(
+                name=sm_name,
+                definition=definition_str,
+                roleArn=snf_role_arn,
+            )
+
+            sfn_snapshot.add_transformer(sfn_snapshot.transform.sfn_sm_create_arn(creation_resp, i))
+
+            state_machine_arn: str = creation_resp["stateMachineArn"]
+            state_machine_arns.append(state_machine_arn)
+
+        def _list_state_machines(expected_results_count: int, **kwargs):
+            """Returns a filtered list of relevant State Machines"""
+            state_machines = aws_client.stepfunctions.list_state_machines(**kwargs)
+            filtered_sms = [sm for sm in state_machines["stateMachines"] if sm["name"] in sm_names]
+
+            assert len(filtered_sms) == expected_results_count
+            return filtered_sms
+
+        # expect all state machines to be present
+        wait_until(lambda: _list_state_machines(expected_results_count=13), max_retries=20)
+
+        paginator = aws_client.stepfunctions.get_paginator("list_state_machines")
+        page_iterator = paginator.paginate(maxResults=5)
+
+        # Paginates across all results and filters out any StateMachines not relevant to the test
+        def _verify_paginate_results() -> list:
+            filtered_state_machines = []
+            for page in page_iterator:
+                assert 0 < len(page["stateMachines"]) <= 5
+
+                filtered_page = [sm for sm in page["stateMachines"] if sm["name"] in sm_names]
+                if filtered_page:
+                    sm_name_set = {sm.get("name") for sm in filtered_state_machines}
+                    # assert that none of the State Machines being added are already present
+                    assert not any(sm.get("name") in sm_name_set for sm in filtered_page)
+
+                    filtered_state_machines.extend(filtered_page)
+
+            assert len(filtered_state_machines) == len(sm_names)
+            return filtered_state_machines
+
+        # Since ListStateMachines is eventually consistent, we should re-attempt pagination
+        listed_state_machines = retry(_verify_paginate_results, retries=20, sleep=1)
+        sfn_snapshot.match("list-state-machines-page-2", listed_state_machines[-3:])
```

Comment:
Is pagination fully reproducible on AWS? (i.e., guaranteed ordering by name according to our implementation)

Reply:
I believe so. Parity tests were initially failing when the provider was sorting state machines by `creationDate`. Here, page 2 consisted of  `state_machine_{10:12}` (in that order). 

Equivalent page 2 when run against AWS returned `state_machine_{7:9}` (due to alphanumeric ordering). I attached the snippet below:

https://github.com/localstack/localstack/blob/e813ac5b77c72af79b00c2a02d8e6dfd8fc7a86d/tests/aws/services/stepfunctions/v2/test_sfn_api.snapshot.json#L1775-L1793